### ISSUE TITLE
compile_commands.json has invalid `arguments` when passing defines containing quoted strings 

### ISSUE THREAD

**Issue by jorgenpt** [fastbuild/fastbuild#1022]
Per #873, we're passing a couple of preprocessor defined strings like this:

```
                #if __WINDOWS__
                    .CompilerOptions + ''
                        + ' -DYAML_VERSION_STRING=\"0.2.5\"'
                #else
                    .CompilerOptions + ''
                        + ' -DYAML_VERSION_STRING="0.2.5"'
                #endif
```

This produces bad (and different) output for both Linux and Windows. On Windows, it produces:

```
"arguments": [..., "-DYAML_VERSION_STRING=\\0.2.5\\"]
```

On Linux, it produces:

```
"arguments": [..., "-DYAML_VERSION_STRING=0.2.5"]
```

Neither of these are accurate and running them will not reproduce the results of building with those CompilerOptions in FASTBuild.

**Comment by ffulin** [comment]
There are a number of things in play here:
 - AString::Tokenize (the core string tokenization function) has assumptions about quotes and that's used as part of tokenizing the args for compile_commands (seems like this is the cause of the missing quotes)
 - Process spawning has differences in behavior, requiring the #if work-around in your example

In addition to the issue you linked, there are some other related issues, like: https://github.com/fastbuild/fastbuild/issues/59, https://github.com/fastbuild/fastbuild/issues/510, https://github.com/fastbuild/fastbuild/issues/741 (and perhaps others)

The ideal long-term outcome would be to resolve the inconsistency that requires #if switches in the bff and to then resolve the issues unique to compile_commands. It may be possible to incrementally work towards that, but it might also require several fixes at the same time due to the nature of how these things interact. In any case, some more investigation will be needed.

**Comment by ffulin** [comment]
Various arg parsing fixes have been implemented in https://github.com/fastbuild/fastbuild/commit/36a8b4dbce16c930afcd14210c80a69529ec3ada

This does not directly fix the behavior of compile_commands.json reported in this issue, but:
a) removes the need for the #if in your define example (the Windows version works on all platforms)
b) adds the necessary functionality in code to allow the compile_commands to be fixed properly now that underlying command line arg handling is consistent

**Comment by ffulin** [comment]
In addition to the fixes to general arg parsing mentioned above, some additional fixes to arg handling for -compb specifically have now been applied in https://github.com/fastbuild/fastbuild/commit/fe10e5a12bb8ee282425dbb00c447f0a099cce24.

These fixes combined resolve the issue in this thread.

The fixes will be in the next release (v1.13). I'll keep this issue open until that version is released.

**Comment by ffulin** [comment]
v1.13 has now been released and includes these fixes. Thanks again for reporting the bug.
